pandas/core/series.py

```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
             elif isinstance(data, dict):
                 if index is None:
                     if isinstance(data, OrderedDict):
-                        index = Index(data)
+                        original_index = data.keys()
                     else:
-                        index = Index(_try_sort(data))
+                        original_index = _try_sort(data)
+
+                    index = Index(original_index)
+
                 try:
-                    if isinstance(index, DatetimeIndex):
-                        # coerce back to datetime objects for lookup
-                        data = lib.fast_multiget(data, index.astype('O'),
-                                                 default=np.nan)
-                    elif isinstance(index, PeriodIndex):
-                        data = [data.get(i, nan) for i in index]
+                    # lib.fast_multiget raises TypeError if type(data) != dict
+
+                    if lib.infer_dtype(data) == lib.infer_dtype(index.values):
```

Comment:
Here we try to avoid a call to `np.array()` by checking if we can access `data` values using `index.values`. However I checked the code for `infer_dtype` and it may be quite complex. Maybe we should just go straight to the `else`?